prometheus: Add settings to prometheus plugin

yolossn opened the pull request:
This patch adds settings to the prometheus plugin
allowing the users to configure the service
address, timespan, enable metrics for each
cluster.

Fixes:
https://github.com/headlamp-k8s/plugins/issues/24

Includes the fix suggested here:
https://github.com/headlamp-k8s/plugins/issues/37#issuecomment-2230345145

Steps to Test:

1. Run prometheus plugin from the `prom_settings` branch.
2. Go to settings and configure the prometheus plugin, ie choose auto detect or provide service details to reach the prometheus service in the cluster.
3. Check if the configuration is working for the cluster by going to pod/namespace detail view.

illume commented:
Is there an issue to link this to?

illume commented:
With these changes will it work with Azure Monitoring?

illume commented:
This comment was made in this issue: https://github.com/headlamp-k8s/plugins/issues/24#issuecomment-1952884113
> In addition, the message is not very clear: "Install Prometheus for accessing metrics charts" Does this mean installing a Prometheus server or the Prometheus plugin for Headlamp?

Before closing #24 that should addressed IMHO.

illume commented on prometheus/package.json:
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "prometheus",
+  "name": "@headlamp-k8s/prometheus",
```
I think the app/ bundling will need to be changed to get the new plugin name too?

yolossn replied:
Nope, we just provide the build of the plugin [here](https://github.com/headlamp-k8s/headlamp/blob/main/app/app-build-manifest.json) so this shouldn't affect.

yolossn commented on prometheus/src/chart.tsx:
```diff
@@ -3,6 +3,49 @@ import { Box, useTheme } from '@mui/material';
+ * @param {string} interval - The time interval (e.g., '10m', '1h', '24h', 'week').
+ * @returns {Object} An object containing the 'from' timestamp, 'to' timestamp, and 'step' in seconds.
+ */
+export function getTimeRange(interval: string): { from: number; to: number; step: number } {
```
Not direct code but similar to the interval to time range conversions in grafana. 
https://github.com/grafana/grafana/blob/0bd0dc374a86ae5d54905fb665348831ebf8425d/packages/grafana-data/src/datetime/rangeutil.ts

yolossn commented on prometheus/src/chart.tsx:
```diff
@@ -12,6 +55,7 @@ export interface ChartProps {
     dataProcessor: (data: any) => any[];
   }>;
   fetchMetrics: (query: object) => Promise<any>;
+  interval: string;
```
Done

yolossn commented on prometheus/src/settings.tsx:
```diff
@@ -0,0 +1,170 @@
+ * @param {Function} props.onDataChange - The function to call when the data changes.
+ * @returns {JSX.Element} The Settings component.
+ */
+export function Settings(props) {
```
Created the SettingsProps and documented it

yolossn commented on prometheus/src/settings.tsx:
```diff
@@ -0,0 +1,170 @@
+import { NameValueTable } from '@kinvolk/headlamp-plugin/lib/components/common';
```
Refactored this and other components into individual folders.

yolossn commented on prometheus/src/chart.tsx:
```diff
@@ -3,6 +3,49 @@ import { Box, useTheme } from '@mui/material';
+ * @returns {Object} An object containing the 'from' timestamp, 'to' timestamp, and 'step' in seconds.
+ */
+export function getTimeRange(interval: string): { from: number; to: number; step: number } {
+  const now = Math.floor(Date.now() / 1000);
```
added the tests.

yolossn commented:
> With these changes will it work with Azure Monitor? I started an issue for this here:
> 
> * [prometheus: Azure Monitoring integration #89](https://github.com/headlamp-k8s/plugins/issues/89)

Nope this is not addressed in this PR. This PR provides a way for the user to configure the plugin to fetch the prometheus metrics from a service with in the cluster.

> I think these are the related issues for this PR?
> 
> * [Prometheus Plug-In: Also detect victoriametrics #37](https://github.com/headlamp-k8s/plugins/issues/37)
> * [prometheus: make backend configurable for now admin users #24](https://github.com/headlamp-k8s/plugins/issues/24)

Yes these are related, now that we provide a way for the user to configure which service to fetch the prometheus metrics from, they should be able to point it to Victoria Metrics or other prometheus compatible service.

This PR doesn't address the requirement related to letting another plugin set the service for a cluster dynamically. (mentioned in #24).

yolossn commented:
> This comment was made in this issue: [#24 (comment)](https://github.com/headlamp-k8s/plugins/issues/24#issuecomment-1952884113)
> 
> > In addition, the message is not very clear: "Install Prometheus for accessing metrics charts" Does this mean installing a Prometheus server or the Prometheus plugin for Headlamp?
> 
> Before closing #24 that should addressed IMHO.

I have updated the message and added a link to the plugin config in the banner itself. If you think the update message is still not clear, please suggest one.

illume reviewed (approved):
👍 thanks folks
